File: packages/shared/src/expense-category.ts
import { coerceCategory } from "./category-coercion";
import type { Category } from "./category-definitions";

/**
 * Category for an expense write:
 * - consults note words when the model emits the vague "Other"
 * - prevents expenses from being stored under Income
 */
export function coerceExpenseCategory(
  value: string | null | undefined,
  note: string | null | undefined,
): Category {
  const cat = coerceCategory(value);
  if (cat === "Income") return categoryFromExpenseNote(note);
  // "Other" -> try to salvage a real category from the note (categoryFromExpenseNote already
  // returns "Other" when nothing matches, so no extra ternary is needed).
  if (cat === "Other") return categoryFromExpenseNote(note);
  return cat;
}

function categoryFromExpenseNote(note: string | null | undefined): Category {
  if (!note) return "Other";
  for (const word of note.toLowerCase().split(/[^a-z0-9]+/)) {
    if (!word) continue;
    // Route each word through coerceCategory (not just CATEGORY_SYNONYMS) so a LITERAL category name
    // in the note — "food", "transport", "bills" — is salvaged too, not only slang. An expense note
    // word is never allowed to classify the write as Income.
    const hit = coerceCategory(word);
    if (hit !== "Other" && hit !== "Income") return hit;
  }
  return "Other";
}

File: crates/shared/src/expense_category.rs
use crate::categories::{coerce_category, Category};

#[must_use]
pub fn coerce_expense_category(value: Option<&str>, note: Option<&str>) -> Category {
    match value.map(coerce_category).unwrap_or(Category::Other) {
        Category::Income | Category::Other => category_from_expense_note(note),
        category => category,
    }
}

fn category_from_expense_note(note: Option<&str>) -> Category {
    let Some(note) = note else {
        return Category::Other;
    };
    for word in note
        .to_ascii_lowercase()
        .split(|ch: char| !ch.is_ascii_alphanumeric())
    {
        if word.is_empty() {
            continue;
        }
        let hit = coerce_category(word);
        if !matches!(hit, Category::Other | Category::Income) {
            return hit;
        }
    }
    Category::Other
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn prevents_expenses_from_becoming_income() {
        assert_eq!(
            coerce_expense_category(Some("Income"), Some("grab")),
            Category::Transport
        );
        assert_eq!(
            coerce_expense_category(Some("Other"), Some("food")),
            Category::Food
        );
        assert_eq!(
            coerce_expense_category(Some("Bills"), None),
            Category::Bills
        );
    }
}
